### PR TITLE
Make `PeriodicallyFlushingMemoryHandler` (buffered logging handler which periodically flushes itself) the default logging handler.

### PR DESCRIPTION
**Pull request by reivilibre**
merge blocked:
- [x] check that #10517 is successful
- [x] check that it doesn't explode matrix.org

### CODE REVIEW

File: changelog.d/10518.feature [hunk 1]
```diff
@@ -0,0 +1 @@
+Make `PeriodicallyFlushingMemoryHandler` (buffered logging handler which periodically flushes itself) the default logging handler.
```
Comment (richvdh) [inline changelog.d/10518.feature:1]:
```suggestion
The default logging handler for new installations is now `PeriodicallyFlushingMemoryHandler`, a buffered logging handler which periodically flushes itself.
```

**Review by richvdh — CHANGES_REQUESTED** [review]
https://github.com/matrix-org/synapse/blob/develop/docker/conf/log.config needs an update too.

File: docs/sample_log_config.yaml [hunk 2]
```diff
@@ -28,14 +28,17 @@ handlers:
     # will be a delay for INFO/DEBUG logs to get written, but WARNING/ERROR
```
Comment (richvdh) [inline docs/sample_log_config.yaml:28]: this isn't quite accurate any more.

**Review by richvdh — APPROVED** [review]
looks fine, though given that `sample_log_config` is the only place that this thing is documented, I wonder if we could do a bit better with it. How do `capacity`, `flushLevel` and `period` interact? What are their default values if left unset?

File: docs/sample_log_config.yaml (hunk 2 above)
Comment (reivilibre) [reply docs/sample_log_config.yaml]: I see what you mean, but it also *kind* of is. I have updated the wording now, mind letting me know if you think it's better?

File: synapse/config/logger.py [hunk 3]
```diff
@@ -67,18 +67,27 @@
...
+    # WARNING/ERROR logs will still be flushed immediately, but there will be a
+    # delay (of up to `period` seconds, or until the buffer is full with
+    # `capacity` messages) before INFO/DEBUG logs get written.
     buffer:
-        class: logging.handlers.MemoryHandler
+        class: synapse.logging.handlers.PeriodicallyFlushingMemoryHandler
         target: file
-        # The capacity is the number of log lines that are buffered before
-        # being written to disk. Increasing this will lead to better
+        # The capacity is the maximum number of log lines that are buffered
```
Comment (richvdh) [inline synapse/config/logger.py:77]: probably worth adding some whitespace between these options:
```suggestion

        # The capacity is the maximum number of log lines that are buffered
```

File: synapse/config/logger.py [hunk 4]
```diff
@@ -67,18 +67,27 @@
...
         capacity: 10
-        flushLevel: 30  # Flush for WARNING logs as well
+        # Logs with a level at or above the flush level will cause the buffer to
+        # be flushed immediately.
+        # Default value: 40 (ERROR)
+        # Other values: 50 (CRITICAL), 30 (WARNING), 20 (INFO), 10 (DEBUG)
+        flushLevel: 30  # Flush immediately for WARNING logs and higher
+        # The period of time, in seconds, between forced flushes.
+        # Messages will not be delayed for longer than this time.
+        period: 5
```
Comment (richvdh) [inline synapse/config/logger.py:90]: what's the default for this? what happens if it is omitted?

File: docker/conf/log.config [hunk 5]
```diff
@@ -18,18 +18,27 @@ handlers:
...
-    flushLevel: 30  # Flush for WARNING logs as well
+    # Logs with a level at or above the flush level will cause the buffer to
+    # be flushed immediately.
+    # Default value: 40 (ERROR)
+    # Other values: 50 (CRITICAL), 30 (WARNING), 20 (INFO), 10 (DEBUG)
+    flushLevel: 30  # Flush immediately for WARNING logs and higher
+    # The period of time, in seconds, between forced flushes.
+    # Messages will not be delayed for longer than this time.
+    period: 5
 {% endif %}
```
Comment (richvdh) [inline docker/conf/log.config:42]: I'd actually be quite happy for this to be very basic, and just link to the sample_log_config for docs. But this is fine too.

**Review by richvdh — COMMENTED** [review]
much better! A couple of little nits.

**Review by richvdh — APPROVED** [review]

**Pull request merged**